File: apps/native/Sources/GozdCore/ProjectKey.swift
import CryptoKit
import Foundation

// プロジェクト固有の永続化先を決める projectKey の生成。
//
// projectKey は `<repoName>-<sha256(realpath)[0..12]>`。
// 同じ main repo の worktree 配下からどの dir で呼んでも、main repo root に解決した上で
// 同一 projectKey が出るようにする。
//
// 利用箇所:
// - `~/.config/gozd/projects/<projectKey>/config.json`（ProjectConfigStore）
// - `~/.config/gozd/projects/<projectKey>/tasks.json`（TaskStore）
// - `~/.local/share/gozd/worktrees/<projectKey>/<leaf>`（WorktreeOps）
//
// 形式変更は全 store の保存先を変えるため、変更時は移行コードを別途用意する必要がある。
public enum ProjectKey {
  /// すでに main repo root と分かっている dir から projectKey を生成する。
  /// 内部で realpath（symlink 解決）してからハッシュする。
  public static func compute(forMainRepoRoot dir: String) -> String {
    let resolved = (dir as NSString).resolvingSymlinksInPath
    let repoName = (resolved as NSString).lastPathComponent
    let digest = SHA256.hash(data: Data(resolved.utf8))
    let hash = digest.compactMap { String(format: "%02x", $0) }.joined()
    return "\(repoName)-\(String(hash.prefix(12)))"
  }

  /// 任意の dir（worktree 配下を含む）から main repo root を引いて projectKey を生成する。
  /// `git rev-parse --git-common-dir` の親が main worktree のパス。git 外 / 失敗時は dir 自体を使う。
  public static func resolveAndCompute(for dir: String) -> String {
    return compute(forMainRepoRoot: resolveMainRepoRoot(for: dir))
  }

  /// `git rev-parse --git-common-dir` の出力（典型的には `<main-worktree>/.git`）の親を realpath。
  /// 失敗時は dir 自体を realpath して返す（git 外でも壊れないため）。
  public static func resolveMainRepoRoot(for dir: String) -> String {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
    process.arguments = ["git", "rev-parse", "--git-common-dir"]
    process.currentDirectoryURL = URL(fileURLWithPath: dir)
    process.environment = ProcessInfo.processInfo.environment
    let stdoutPipe = Pipe()
    let stderrPipe = Pipe()
    process.standardOutput = stdoutPipe
    process.standardError = stderrPipe
    do {
      try process.run()
      process.waitUntilExit()
      let data = stdoutPipe.fileHandleForReading.readDataToEndOfFile()
      _ = stderrPipe.fileHandleForReading.readDataToEndOfFile()
      if process.terminationStatus != 0 {
        return (dir as NSString).resolvingSymlinksInPath
      }
      let text = String(decoding: data, as: UTF8.self)
        .trimmingCharacters(in: .whitespacesAndNewlines)
      // common-dir が相対パスなら dir 起点で resolve する
      let commonDir =
        text.hasPrefix("/")
        ? text
        : (URL(fileURLWithPath: dir).appendingPathComponent(text)).path
      let parent = (commonDir as NSString).deletingLastPathComponent
      return (parent as NSString).resolvingSymlinksInPath
    } catch {
      return (dir as NSString).resolvingSymlinksInPath
    }
  }
}

File: apps/electron/src/projectKey.ts
// repo の識別子（projectKey）と、gozd が worktree を置く場所の解決。
//
// - projectKey は `<repoName>-<sha256(realpath)[0..12]>`。worktree 配下のどの dir から
//   呼ばれても main repo root に解決した上で同一 projectKey に揃える
// - worktree の配置先と、プロジェクト単位の永続ファイルの置き場所が同じ値を共有する

import { tryCatch } from "@gozd/shared";
import { createHash } from "node:crypto";
import { realpathSync } from "node:fs";
import { homedir } from "node:os";
import { basename, dirname, isAbsolute, join } from "node:path";
import { runGit } from "./git/gitRunner";

/** realpath 解決。対象が存在しない等で失敗したら入力をそのまま返す */
function realpathOrSelf(dir: string): string {
  const result = tryCatch(() => realpathSync(dir));
  return result.ok ? result.value : dir;
}

/** `git rev-parse --git-common-dir` の親 = main worktree のパス。git 外 / 失敗時は dir 自体 */
export async function resolveMainRepoRoot(dir: string): Promise<string> {
  const result = await tryCatch(runGit(["rev-parse", "--git-common-dir"], dir));
  if (!result.ok) return realpathOrSelf(dir);
  const text = result.value.trim();
  // common-dir が相対パスなら dir 起点で resolve する
  const commonDir = isAbsolute(text) ? text : join(dir, text);
  return realpathOrSelf(dirname(commonDir));
}

/** main repo root から projectKey を生成する。形式変更は全 store の保存先を変えるため
 * 変更時は移行コードが必要 */
function computeProjectKey(mainRepoRoot: string): string {
  const resolved = realpathOrSelf(mainRepoRoot);
  const hash = createHash("sha256").update(resolved, "utf8").digest("hex");
  return `${basename(resolved)}-${hash.slice(0, 12)}`;
}

/** dir（main / worktree / 配下 subdir のいずれでも可）から projectKey を解決する。
 * worktree 配置先の決定（worktreeOps）と永続ファイルパスの両方がこの値を共有する */
export async function resolveProjectKey(dir: string): Promise<string> {
  return computeProjectKey(await resolveMainRepoRoot(dir));
}

/** gozd の worktree 配置 root（projectKey 抜き）。`<root>/<projectKey>/<leaf>` が各 worktree のパス。
 * worktreeOps（worktree の作成先）と claudeSessionLog（revive の cwd prefix 判定）が同一 base を
 * 指すことに revive の「cwd 1 バイト一致」が依存するため、literal を 2 箇所に散らさず SSOT を置く。 */
export function gozdWorktreesRoot(): string {
  return join(homedir(), ".local", "share", "gozd", "worktrees");
}
